MyPy Error Fix | Open to all

As from #294 I'd like to slowly start decreasing the amount of MyPy errors throughout the project.

Any PRs are welcome to slowly remove, even just one mypy error at a time. Make sure in the PR not to add "closes 319" but just to add "Related issue".

Open to all, thanks @introkun for the idea and @Domejko for making a start.

Current (11/04/2024):
```
$ mypy .
backend\api\admin\api_keys.py:1: error: Skipping analyzing "django.contrib": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\admin\api_keys.py:1: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
backend\api\admin\api_keys.py:2: error: Skipping analyzing "django.http": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\admin\api_keys.py:3: error: Skipping analyzing "django.shortcuts": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\admin\urls.py:1: error: Skipping analyzing "django.urls": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\base\modal.py:3: error: Skipping analyzing "django.contrib": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\base\modal.py:4: error: Skipping analyzing "django.http": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\base\modal.py:6: error: Skipping analyzing "django.shortcuts": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\base\notifications.py:1: error: Skipping analyzing "django.http": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\base\notifications.py:2: error: Skipping analyzing "django.shortcuts": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\base\urls.py:1: error: Skipping analyzing "django.urls": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\clients\fetch.py:1: error: Skipping analyzing "django.db.models": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\clients\fetch.py:2: error: Skipping analyzing "django.http": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\clients\fetch.py:3: error: Skipping analyzing "django.shortcuts": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\clients\fetch.py:4: error: Skipping analyzing "django.views.decorators.http": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\clients\urls.py:1: error: Skipping analyzing "django.urls": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\currency_converter\convert.py:3: error: Skipping analyzing "django.contrib": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\currency_converter\convert.py:4: error: Skipping analyzing "django.http": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\currency_converter\convert.py:5: error: Skipping analyzing "django.shortcuts": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\currency_converter\convert.py:6: error: Skipping analyzing "forex_python.converter": module is installed, but missing library stubs or py.typed marker  [import-untyped]        
backend\api\currency_converter\urls.py:1: error: Skipping analyzing "django.urls": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\emails\fetch.py:1: error: Skipping analyzing "django.contrib": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\emails\fetch.py:2: error: Skipping analyzing "django.core.paginator": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\emails\fetch.py:3: error: Skipping analyzing "django.db.models": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\emails\fetch.py:4: error: Skipping analyzing "django.http": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\emails\fetch.py:5: error: Skipping analyzing "django.shortcuts": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\emails\fetch.py:6: error: Skipping analyzing "django_ratelimit.core": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\emails\fetch.py:7: error: Skipping analyzing "django_ratelimit.decorators": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\emails\send.py:9: error: Skipping analyzing "django.contrib": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\emails\send.py:10: error: Skipping analyzing "django.core.exceptions": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\emails\send.py:11: error: Skipping analyzing "django.core.handlers.wsgi": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\emails\send.py:12: error: Skipping analyzing "django.core.validators": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\emails\send.py:13: error: Skipping analyzing "django.db.models": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\emails\send.py:14: error: Skipping analyzing "django.http": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\emails\send.py:15: error: Skipping analyzing "django.shortcuts": module is installed, but missing library stubs or py.typed marker  [import-untyped]
backend\api\emails\send.py:16: error: Skipping analyzing "django.views.decorators.http": module is installed, but missing library stubs or py.typed marker  [import-untyped]
settings\helpers.py: error: Source file found twice under different module names: "helpers" and "settings.helpers"
settings\helpers.py: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#mapping-file-paths-to-modules for more info
settings\helpers.py: note: Common resolutions include: a) adding `__init__.py` somewhere, b) using `--explicit-package-bases` or adjusting MYPYPATH
Found 36 errors in 12 files (errors prevented further checking)
```

--- a/backend/api/invoices/create/services/add.py
+++ b/backend/api/invoices/create/services/add.py
@@ -9,7 +9,7 @@ from backend.types.htmx import HtmxHttpRequest
 @require_http_methods(["POST"])
 def add_service(request: HtmxHttpRequest):
     context: dict = {}
-    existing_service = request.POST.get("existing_service", "")
+    existing_service = request.POST.get("existing_service", 0)
 
     try:
         existing_service_obj = InvoiceProduct.objects.get(user=request.user, id=existing_service)

--- a/backend/models.py
+++ b/backend/models.py
@@ -2,8 +2,7 @@ from __future__ import annotations
 
 from datetime import datetime
 from decimal import Decimal
-from typing import Literal
-from typing import NoReturn
+from typing import Literal, Union
 from uuid import uuid4
 
 from django.contrib.auth.hashers import check_password
@@ -358,7 +357,7 @@ class Invoice(models.Model):
             return self.payment_status
 
     @property
-    def get_to_details(self) -> tuple[str, dict[str, str]] | tuple[str, Client]:
+    def get_to_details(self) -> tuple[str, dict[str, str | None]] | tuple[str, Client]:
         """
         Returns the client details for the invoice
         "client" and Client object if client_to
@@ -378,13 +377,13 @@ class Invoice(models.Model):
             subtotal += item.get_total_price()
         return Decimal(round(subtotal, 2))
 
-    def get_tax(self, amount: Decimal = 0.00) -> Decimal:
-        amount = amount or self.get_subtotal()
+    def get_tax(self, amount: Decimal = Decimal(0.00)) -> Decimal:
+        amount = amount or Decimal(self.get_subtotal())
         if self.vat_number:
             return Decimal(round(amount * Decimal(0.2), 2))
         return Decimal(0)
 
-    def get_percentage_amount(self, subtotal: float = 0.00) -> Decimal:
+    def get_percentage_amount(self, subtotal: Decimal = Decimal(0.00)) -> Decimal:
         total = subtotal or self.get_subtotal()
 
         if self.discount_percentage > 0:
@@ -401,7 +400,7 @@ class Invoice(models.Model):
         total -= discount_amount
 
         if 0 > total:
-            total = 0
+            total = Decimal(0)
         else:
             total -= self.get_tax(total)
 
@@ -649,6 +648,7 @@ class QuotaLimit(models.Model):
         return self.name
 
     def get_quota_limit(self, user: User, quota_limit: QuotaLimit | None = None):
+        user_quota_override: QuotaOverrides | QuotaLimit
         try:
             if quota_limit:
                 user_quota_override = quota_limit
@@ -669,6 +669,8 @@ class QuotaLimit(models.Model):
             return "Not available"
 
     def strict_goes_above_limit(self, user: User, extra: str | int | None = None, add: int = 0) -> bool:
+        current: Union[int, None, QuerySet[QuotaUsage], Literal['Not Available']]
+
         current = self.strict_get_quotas(user, extra)
         current = current.count() if current != "Not Available" else None
         return current + add >= self.get_quota_limit(user) if current else False
@@ -681,35 +683,38 @@ class QuotaLimit(models.Model):
         :return: QuerySet of quota usages OR "Not Available" if utilisation isn't available (e.g. per invoice you can't get in total)
         """
         current = None
-        quota_limit = quota_limit.quota_usage if quota_limit else QuotaUsage.objects.filter(user=user, quota_limit=self)
+        if quota_limit is not None:
+            quota_lim = quota_limit.quota_usage
+        else:
+            quota_lim = QuotaUsage.objects.filter(user=user, quota_limit=self)  # type: ignore[assignment]
 
         if self.limit_type == "forever":
             current = self.quota_usage.filter(user=user, quota_limit=self)
         elif self.limit_type == "per_month":
             current_month = timezone.now().month
             current_year = timezone.now().year
-            current = quota_limit.filter(created_at__year=current_year, created_at__month=current_month)
+            current = quota_lim.filter(created_at__year=current_year, created_at__month=current_month)
         elif self.limit_type == "per_day":
             current_day = timezone.now().day
             current_month = timezone.now().month
             current_year = timezone.now().year
-            current = quota_limit.filter(created_at__year=current_year, created_at__month=current_month, created_at__day=current_day)
+            current = quota_lim.filter(created_at__year=current_year, created_at__month=current_month, created_at__day=current_day)
         elif self.limit_type in ["per_client", "per_invoice", "per_team", "per_receipt", "per_quota"] and extra:
-            current = quota_limit.filter(extra_data=extra)
+            current = quota_lim.filter(extra_data=extra)
         else:
             return "Not Available"
         return current
 
     @classmethod
-    def delete_quota_usage(cls, quota_limit: str | QuotaLimit, user: User, extra, timestamp=None) -> NoReturn:
+    def delete_quota_usage(cls, quota_limit: str | QuotaLimit, user: User, extra, timestamp=None):
         quota_limit = cls.objects.get(slug=quota_limit) if isinstance(quota_limit, str) else quota_limit
 
         all_usages = quota_limit.strict_get_quotas(user, extra)
         closest_obj = None
 
         if all_usages.count() > 1 and timestamp:
-            earliest: QuotaUsage = all_usages.filter(created_at__gte=timestamp).order_by("created_at").first()
-            latest: QuotaUsage = all_usages.filter(created_at__lte=timestamp).order_by("created_at").last()
+            earliest: QuotaUsage | None = all_usages.filter(created_at__gte=timestamp).order_by("created_at").first()
+            latest: QuotaUsage | None = all_usages.filter(created_at__lte=timestamp).order_by("created_at").last()
 
             if earliest and latest:
                 time_until_soonest_obj = abs(earliest.created_at - timestamp)

--- a/backend/models.py
+++ b/backend/models.py
@@ -378,7 +378,7 @@ class Invoice(models.Model):
         return Decimal(round(subtotal, 2))
 
     def get_tax(self, amount: Decimal = Decimal(0.00)) -> Decimal:
-        amount = amount or Decimal(self.get_subtotal())
+        amount = amount or self.get_subtotal()
         if self.vat_number:
             return Decimal(round(amount * Decimal(0.2), 2))
         return Decimal(0)

--- a/backend/models.py
+++ b/backend/models.py
@@ -1,5 +1,6 @@
 from __future__ import annotations
 
+import typing
 from datetime import datetime
 from decimal import Decimal
 from typing import Literal, Union
@@ -706,6 +707,7 @@ class QuotaLimit(models.Model):
         return current
 
     @classmethod
+    @typing.no_type_check
     def delete_quota_usage(cls, quota_limit: str | QuotaLimit, user: User, extra, timestamp=None):
         quota_limit = cls.objects.get(slug=quota_limit) if isinstance(quota_limit, str) else quota_limit
 

--- a/backend/models.py
+++ b/backend/models.py
@@ -669,7 +669,7 @@ class QuotaLimit(models.Model):
             return "Not available"
 
     def strict_goes_above_limit(self, user: User, extra: str | int | None = None, add: int = 0) -> bool:
-        current: Union[int, None, QuerySet[QuotaUsage], Literal['Not Available']]
+        current: Union[int, None, QuerySet[QuotaUsage], Literal["Not Available"]]
 
         current = self.strict_get_quotas(user, extra)
         current = current.count() if current != "Not Available" else None

--- a/backend/views/core/__init__.py
+++ b/backend/views/core/__init__.py
@@ -3,5 +3,5 @@ from .other import index, errors
 from .auth import login
 from .settings import view as settings_view, teams
 from .invoices import dashboard as invoices_dashboard, create, view as invoices_view, manage_access
-from .clients import dashboard as clients_dashboard, create
+from .clients import dashboard as clients_dashboard, create as create_client
 from .receipts import dashboard as receipts_dashboard

--- a/backend/views/core/auth/login.py
+++ b/backend/views/core/auth/login.py
@@ -75,7 +75,7 @@ def login_manual(request: HtmxAnyHttpRequest):  # HTMX POST
         messages.error(request, "Incorrect email or password")
         return render_toast_message(request)
 
-    if user.awaiting_email_verification and ARE_EMAILS_ENABLED:  # type: ignore[union-attr]
+    if user.awaiting_email_verification and ARE_EMAILS_ENABLED:  # type: ignore[attr-defined]
         messages.error(request, "You must verify your email before logging in.")
         return render_toast_message(request)
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -122,6 +122,19 @@ plugins = [
     "mypy_django_plugin.main",
     "mypy_drf_plugin.main"
 ]
+disable_error_code = ["annotation-unchecked"]
+explicit_package_bases = true
+incremental = false
+exclude = [
+    "venv",
+    ".venv",
+    "env",
+    "ENV",
+    "ENV_DIR",
+    "env.bak",
+    "venv.bak",
+    "media"
+]
 
 [tool.django-stubs]
 django_settings_module = "settings.settings"

